custom-targets/vertex-ai/model-deployer/deploy.go

```diff
@@ -0,0 +1,304 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to download deploy input from %s: %v", d.req.ManifestGCSPath, err)
+	}
+
+	fmt.Printf("Downloaded deploy input manifest.\n")
+
+	manifestData, err := d.applyModel(ctx, localManifest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to deploy model: %v", err)
+	}
+
+	mURI, err := d.req.UploadArtifact(ctx, d.gcsClient, "manifest.yaml", &clouddeploy.GCSUploadContent{Data: manifestData})
+	if err != nil {
+		return nil, fmt.Errorf("error uploading deploy artifact: %v", err)
+	}
+
+	return &clouddeploy.DeployResult{
+		ResultStatus:  clouddeploy.DeploySucceeded,
+		ArtifactFiles: []string{mURI},
+	}, nil
+}
+
+func (d *deployer) addCommonMetadata(rs *clouddeploy.DeployResult) {
+	if rs.Metadata == nil {
+		rs.Metadata = map[string]string{}
+	}
+	rs.Metadata[clouddeploy.CustomTargetSourceMetadataKey] = aiDeployerSampleName
+	rs.Metadata[clouddeploy.CustomTargetSourceSHAMetadataKey] = clouddeploy.GitCommit
+}
+
+func deployModelFromManifest(path string) (*aiplatform.GoogleCloudAiplatformV1DeployModelRequest, error) {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("error reading manifest file: %v", err)
+	}
+
+	deployModelRequest := &aiplatform.GoogleCloudAiplatformV1DeployModelRequest{}
+
+	if err = yaml.Unmarshal(data, deployModelRequest); err != nil {
+		return nil, fmt.Errorf("unable to parse deploy model deployModelRequest from manifest file: %v", err)
+	}
+
+	return deployModelRequest, nil
+}
+
+func (d *deployer) applyModel(ctx context.Context, localManifest string) ([]byte, error) {
+
+	deployModelRequest, err := deployModelFromManifest(localManifest)
+	if err != nil {
+		return nil, err
+	}
+
+	endpointRegion, err := fetchRegionFromEndpoint(d.params.endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse region from endpoint resource name: %v", err)
+	}
+
+	aiplatformService, err := newService(ctx, endpointRegion)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create service: %v", err)
+	}
+
+	if d.req.Percentage != 100 {
+		previousModel, err := determinePreviousModel(aiplatformService, d.params.endpoint, deployModelRequest.DeployedModel.Model)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get previous model to canary against: %v", err)
+		}
+		previousPercentage, ok := deployModelRequest.TrafficSplit["previous-model"]
+		if !ok {
+			return nil, fmt.Errorf("expected input manifest trafficSplit stanza to have a 'previous-model' entry but did not find it")
+		}
+		delete(deployModelRequest.TrafficSplit, "previous-model")
+		deployModelRequest.TrafficSplit[previousModel] = previousPercentage
+	}
+	op, err := aiplatformService.Projects.Locations.Endpoints.DeployModel(d.params.endpoint, deployModelRequest).Do()
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to deploy model: %v", err)
+	}
+
+	if err := Poll(ctx, aiplatformService, op); err != nil {
+		return nil, err
+	}
```

Comment:
Suggest moving logic for interacting with the AI service into separate file so the deploy file is a bit more readable. It would remove things like the operation polling logic from here, which isn't necessary for understanding the deploy process. See IM sample in my private repo as a reference.